contrib/bpftrace/taskqlatency.sh

```diff
@@ -0,0 +1,64 @@
+#!/bin/bash
+
+read -r -d '' PROGRAM <<'EOF'
+#include <sys/taskq.h>
+
+kprobe:trace_zfs_taskq_ent__birth
+{
+        $tqent = (struct taskq_ent *)arg0;
+
+        $tqent_id = $tqent->tqent_id;
+        $tq_name = str($tqent->tqent_taskq->tq_name);
+
+        @birth[$tq_name, $tqent_id] = nsecs;
+}
+
+kprobe:trace_zfs_taskq_ent__start
+{
+        $tqent = (struct taskq_ent *)arg0;
+
+        @tqent_id[tid] = $tqent->tqent_id;
+        @tq_name[tid] = str($tqent->tqent_taskq->tq_name);
+
+        @start[@tq_name[tid], @tqent_id[tid]] = nsecs;
+}
+
+kprobe:trace_zfs_taskq_ent__start
+/ @birth[@tq_name[tid], @tqent_id[tid]] /
+{
+        @queue_lat_us[@tq_name[tid]] =
+                hist((nsecs - @birth[@tq_name[tid], @tqent_id[tid]])/1000);
+        delete(@birth[@tq_name[tid], @tqent_id[tid]]);
+}
+
+kprobe:trace_zfs_taskq_ent__finish
+/ @start[@tq_name[tid], @tqent_id[tid]] /
+{
+        $tqent = (struct taskq_ent *)arg0;
+
+        @exec_lat_us[@tq_name[tid], ksym($tqent->tqent_func)] =
+                hist((nsecs - @start[@tq_name[tid], @tqent_id[tid]])/1000);
+        delete(@start[@tq_name[tid], @tqent_id[tid]]);
+}
+
+kprobe:trace_zfs_taskq_ent__finish
+{
+        delete(@tq_name[tid]);
+        delete(@tqent_id[tid]);
+}
+
+END
+{
+        clear(@birth);
+        clear(@start);
+
+        clear(@tq_name);
+        clear(@tqent_id);
+}
+EOF
+
+KVER=$(uname -r)
```

Comment:
Honestly, I'd rather just update the script to work correctly for all platforms, rather than writing a comment to explain why it may be broken on any specific platform(s).

Can you give me some pointers on how I should convert this to a `.in` file, so it'll work on all platforms? e.g. Is there some kind of "include directory" variable that I can use, in place of `/usr/src.../` here?

If there isn't a clean way to do this for all platforms, I'll add the comment; but I'd rather invest the time to fix this properly now, than kick that can down the road, and have to address it later.

I'm not familiar with how to do this via the `.in` files, though, so a little help would be appreciated.

Reply:
@behlendorf would it better to just remove all the bash/shell logic, and let this be a "raw" bpftrace script instead? that'd be most simple (from an implementation standpoint), but would push the requirements of how to specify `-I` and `-include` to the user.. so far, I'm leaning towards doing that.. what do you think?

Reply:
In that case, I'd suggest taking a look at the `etc/systemd/system/Makefile.am`.  It provides a good example of how  to setup the `Makefile.am` rules to perform the needed substitution for a `taskqlatency.sh.in` file. 

If you take a look at the bottom of the `include/sys/Makefile.am` you can see where `make install` intends to place the generic headers.  These aren't tired to any specific kernel.

```
kerneldir = @prefix@/src/zfs-$(VERSION)/include/sys
```

The top-level `Makefile.am` includes the install location for any kernel or build specific object files.  Though you probably want to continue using $(uname -r) for the `$(LINUX_VERSION)` so it works for the running kernel

```
kerneldir = $(prefix)/src/zfs-$(VERSION)/$(LINUX_VERSION)
```

You'll want to substitute those paths in to the script since that's where the build system intends to install them required headers.

Reply:
Before I look into implementing this.. do you prefer having the "wrapper script" that contains the bpftrace script, and the call to bpftrace (with the proper options), like I originally had.. or just the "raw" bpftrace script, without the call to bpftrace, like I currently have in this PR (I just updated it)?

Reply:
At first I liked the wrapper script, but that's almost certainly because I'm not very familiar with `bpftrace`.  There's a lot of appeal in it just-working as a debugging tool without needing to understand anything about `bpftrace`.

That said, I expect we're going to want to start accumulating these useful scripts.  I think the most practical first step would be to include a small `README` in the directory explaining how to invoke them.  Adding per-distribution instructions would be easy.

Then follow up (in this PR or another) with a new `zfs-trace <bpftrace script>` wrapper script which runs `bpftrace` with the right options for your platform.  That could be generated from a `zfs-trace.in` file with the correct platform paths rather than doing it for each bpftrace script.  That would put in place most of what we need to create a `zfs-trace` package which would install the `zfs-trace` driver, and all of the raw bpftrace scripts to a default location on the system.  Then you could run something like `zfs-trace taskqlatency`.

Reply:
@ahrens suggested a script like `zfs-trace` which you're suggesting too.

For our appliance, since we build `bpftrace` ourselves, we opted to instead modify `bpftrace` to automatically add the correct `-I` options to work with the ZFS headers; see [here](https://github.com/delphix/bpftrace/blob/master/src/main.cpp#L454):
```
    auto zdir = std::string("/usr/src/zfs-") + utsname.release;
    if (is_dir(zdir)) {
      extra_flags.push_back("-include");
      extra_flags.push_back(zdir + "/zfs_config.h");
      extra_flags.push_back("-I" + zdir + "/include");
      extra_flags.push_back("-I" + zdir + "/include/spl");
    }
```

I'm not opposed to adding a `zfs-trace` script in here, but I think it would be best if we could push something upstream to `bpftrace` to automatically add the correct paths. I'm not sure how that community would feel about such a change, but I can try to pursue that approach (in conjunction with adding this `zfs-trace` script, for now) if you agree that'd be best. What do you think?

In the mean time, let me try to write a `README.md` file and the `zfs-trace` script, and add this to the PR here.

Reply:
I think it's worth asking the question of the `bpftrace` developers how they would prefer to handle this kind of tracing of external kmods.  In the meanwhile, it'd be nice to have our own `zfs-trace` script.

Reply:
Thinking some more, if we create a `zfs-trace`, how should the script behave if two versions of the ZFS modules are installed?

If we use the `.in` file approach, that'd mean the script will be tied to the specific version of ZFS that was used when building, but that does not necessarily mean when the script is executed, that same version of ZFS will be running.

For example, if we have a system running 0.8.0, then install 0.8.1 without rebooting and attempt to use the 0.8.1 version of the script, it likely will not do the correct thing; it'd use the 0.8.1 headers (right?), even though the actual system is running 0.8.0.

Reply:
Please see my latest commit, and let me know what you think; I'm not sure if I got the Makefile logic correct or not (it seems to work in my limited testing).

Reply:
That's a good point.  It sounds like what we really want is for the script to use the running kernel and loaded zfs kmods.  Something like this:

```
ZVER=$(cat /sys/module/zfs/version)
KVER=$(uname -r)
```

The Lustre build system has this same issue which looking for the ZFS headers and objects.    We could do something similar, I notice they also handle the dkms header install case which would be good.  See the LB_ZFS macro.

https://git.whamcloud.com/?p=fs/lustre-release.git;a=blob_plain;f=config/lustre-build-zfs.m4;hb=HEAD

Reply:
```
$ cat /sys/module/zfs/version
0.8.0-1
```
I guess I should only use the portion preceding the `-` character?

Reply:
Yes.  The `-` splits the VERSION from the RELEASE, we just want the VERSION since downstream maintainers are free to do whatever they want with the RELEASE field.

Reply:
I don't have a good way to test the DKMS directory, so I didn't include that in my latest version.

Reply:
> I don't know why these differ than when I use "make install".

It must be due to differences is the packaging.  Each distribution / packager has their own guidelines for where things should be installed.  At the end of the day, all we can really do is provide mechanisms for them to specify those locations via `configure` options (i,e -bindir=DIR, --sysconfdir=DIR, etc).  And then substitute the requested paths as needed for the files we install, via .in files.